### PR TITLE
Add NativeAOT to build

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
NativeAOT becomes part of the clr subset.

The NativeAOT runtime and libraries are a clr.nativeaotlibs subset.
The NativeAOT object writer (not built by default) is clr.objwriter subset.
The AOT compiler is part of the existing clr.tools subset.

### CODE REVIEW

File: eng/Subsets.props [hunk 1]
```diff
@@ -242,6 +257,22 @@
     <ProjectToBuild Include="$(CoreClrProjectRoot)tools\dotnet-pgo\dotnet-pgo-pack.proj" Pack="true" BuildInParallel="false" Category="clr" Condition="'$(DotNetBuildFromSource)' != 'true' and '$(RuntimeFlavor)' != 'Mono'"/>
   </ItemGroup>
 
+  <ItemGroup Condition="$(_subset.Contains('+clr.nativeaotlibs+')) and '$(NativeAotSupported)' == 'true'">
```
Comment (marek-safar) [inline eng/Subsets.props:260]: Why are non-coreclr pieces located under coreclr (same applies to tools) ?

Comment (MichalStrehovsky) [reply eng/Subsets.props:260]: NativeAOT is a flavor of CoreCLR

Comment (marek-safar) [reply eng/Subsets.props:260]: So there is no runtime agnostic tool or assembly in NativeAOT right now?

Comment (MichalStrehovsky) [reply eng/Subsets.props:260]: We'll be moving files out as we unify the runtimes. E.g. a lot of src\coreclr\nativeaot\System.Private.Reflection.Core\src is basically src\libraries\System.Reflection.MetadataLoadContext\src but they need to be unified as we figure out the right shape of a common reflection stack (also shared with Mono). There's also bits and pieces in NativeAOT's corelib that we would like to share. Now that we have the things here, it's easier to do such refactorings.

**Comment by MichalStrehovsky** [comment]
Cc @safern @jkoritzinsky for general build-infra. This is enabling NativeAOT bits in the CoreCLR build.

File: src/coreclr/nativeaot/Runtime/CommonMacros.h [hunk 2]
```diff
@@ -221,11 +233,11 @@ extern unsigned __int64 g_startupTimelineEvents[NUM_STARTUP_TIMELINE_EVENTS];
 #define C_ASSERT(e) static_assert(e, #e)
 #endif // C_ASSERT
 
-#ifdef __llvm__
+#if defined(__llvm__) || defined(__GNUC__)
```
Comment (kant2002) [inline src/coreclr/nativeaot/Runtime/CommonMacros.h:236]: Isn't this just `#ifndef _MSC_VER` where you want define MSVC specific stuff?

File: eng/native/functions.cmake [hunk 3]
```diff
@@ -367,6 +367,25 @@ function(install_symbol_file symbol_file destination_path)
   endif()
 endfunction()
 
+function(install_static_library targetName destination component)
```
Comment (kant2002) [inline eng/native/functions.cmake:370]: this seems to be part of #62570 and not used anywhere in the build.

Comment (MichalStrehovsky) [reply eng/native/functions.cmake:370]: This is used in all CMakeLists.txt files under nativeaot.

**Comment by MichalStrehovsky** [comment]
@dotnet/runtime-infrastructure this is adding NativeAOT to the CLR build.

`build.cmd clr -rc Debug -lc Release` gets about 15% slower. It's 15% because we are building the NativeAOT bits 3 times on Windows because we need a flavor of the runtime with/without the Server GC, and one more flavor with Server GC and Control Flow Guard enabled. This is to accommodate user-facing configurability options. The overhead is smaller outside Windows because there's no CFG flavor.

**Review by jkotas — APPROVED** [review]

File: eng/Subsets.props [hunk 4]
```diff
@@ -242,6 +257,22 @@
...
+    <ProjectToBuild Include="$(CoreClrProjectRoot)nativeaot\System.Private.CoreLib\src\System.Private.CoreLib.csproj" Category="clr" />
```
Comment (safern) [inline eng/Subsets.props:261]: Why not include these as a wildcard?

**Review by safern — COMMENTED** [review]
Other than a question, LGTM

File: eng/Subsets.props (hunk 4 above)
Comment (MichalStrehovsky) [reply eng/Subsets.props:261]: Good point! #62820

**Pull request merged**